File: Python/653. Two Sum IV - Input is a BST.py
class Solution:
    def findTarget(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: bool
        """
        maple = dict()
        def search(node, k):
            if node is None:
                return False
            if maple.get(k-node.val) is None:
                maple[node.val] = 1
                if search(node.left, k):
                    return True
                if search(node.right, k):
                    return True
                return False
            else:
                return True
        return search(root, k)

File: C++/653. Two Sum IV - Input is a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    unordered_map<int, int> maple;
    bool dfs(TreeNode* node, int k)
    {
        if(node == nullptr) return false;
        if(maple.find(k-node->val)!=maple.end())
            return true;
        maple[node->val] = 1;
        return dfs(node->left, k) || dfs(node->right, k);
    }
    bool findTarget(TreeNode* root, int k) {
        maple.clear();
        return dfs(root,k);
    }
};